Cache ComputedVar

masenf opened the pull request:
Building upon #916 ([👉**view only the diff from 916**](https://github.com/pynecone-io/pynecone/pull/917/files/18d16e4ee246bc6208a2884dbfc1a8d29a02b1e9..53f3232d0539134c1574464f5858fb356f35c45d)), the purpose of this PR is to avoid recalculating potentially expensive `ComputedVar` values unless their dependencies have actually changed.

If many substates depend on an expensive `ComputedVar` in a parent state, then this ensures that it is only recalculated once per dirtying of dependencies.

### All Submissions:

- [x] Have you followed the guidelines stated in [CONTRIBUTING.md](https://github.com/pynecone-io/pynecone/blob/main/CONTRIBUTING.md) file?
- [x] Have you checked to ensure there aren't any other open [Pull Requests](https://github.com/pynecone-io/pynecone/pulls ) for the desired changed?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Type of change

- [x] New feature (non-breaking change which adds functionality)


### New Feature Submission:

- [x] Does your submission pass the tests? 
- [x] Have you linted your code locally prior to submission?

### Changes To Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

masenf committed:
vars: Add ComputedVars.deps function to calculate state Var dependencies

The `deps` function uses the stdlib `dis` module to iterate through the
instructions of the ComputedVar getter function. It keeps track of when
attributes are loaded from "self" and marks those as dependencies. If the
function attempts to call a method of the State, then recursively call `deps`
against the method and update the dependency set with the method's Var
depndencies.

Potentially addresses #870

masenf committed:
state: determine computed_var_dependencies via code examination

Remove the need to runtime-check variable access for each computed var, instead
relying on the new ComputedVar.deps() method to determine the dependencies by
looking at the function's bytecode.

masenf committed:
test_state: assert on computed_var_dependencies

ensure that the computed_var_dependencies are also determined correctly

masenf committed:
test_state: remove unchanged computed vars from delta calculation

Now that state is only including changed vars in the delta, remove any entries
that were not changed.

masenf committed:
test_state: remove `check=True` argument from get_delta

the argument was removed from `State.get_delta` as the code is always checking
for changed ComputedVars.

masenf committed:
var: cache ComputedVar value on instance

save the value of the ComputedVar to avoid recomputing it

if `mark_dirty` is called, then the cached value is deleted and will be
recomputed next time it is needed.

masenf committed:
state: handle dirty ComputedVar

since ComputedVar are now cached they must be marked dirty whenever any of
their dependencies are modified.

process substate deltas first, to avoid extra recalculations of ComputedVar along the way

masenf committed:
test_state: ensure ComputedVar is cached

Only recalculate ComputedVar when dependencies change.

picklelo commented on pynecone/state.py:
```diff
@@ -105,18 +99,10 @@ def __init__(self, *args, parent_state: Optional[State] = None, **kwargs):
-            self.__getattribute__(cvar)
-            self.track_vars = False
-
+        for cvar_name, cvar in self.computed_vars.items():
```
I think we may need to keep the line `self.computed_var_dependencies = defaultdict(set)` - because of Python weirdness when we set the initial value to a mutable type like list/dict it becomes shared by all the subclasses. We ran into this issue a couple times

masenf replied:
_that_ probably explains the weirdness i was hitting w.r.t. accessing that value in a class context vs an instance context. will rework and see if i can add a test for this.

masenf replied:
@picklelo i moved the `defaultdict` assignment back into `__init__`, so that each instance gets a copy, rather than sharing the class copy...

However to make it work, I had to also move the `_init_mutable_values` call _after_ `computed_var_dependencies` is assigned, because `_init_mutable_vars` might call `__setattr__`, which will call `mark_dirty`, which attempts to look up any dirty vars in the `computed_var_dependencies`, which won't work, unless it's been assigned as a default dict... turns out _that_ was the problem I was hitting before.

After moving the `_init_mutable_values` call, there's no problem with assigning `computed_var_dependencies` in `__init__`.

picklelo reviewed (commented):
Awesome! One small comment but looks good initially. I'll review this more thoroughly and try to include in the following release.

masenf committed:
state: initialize computed_var_dependencies in __init__

avoid interactions with subclasses, by overriding the
`computed_var_dependencies` in the instance and assigning dependencies on each.

move _init_mutable_fields after computed_var_dependency determination, since
they may result in `mark_dirty` being called.

masenf committed:
Merge remote-tracking branch 'origin/main' into cache-computed-vars

picklelo reviewed (approved):
Tested this out on `pcweb` and seems to work really well - awesome work! Much cleaner and more performant calculating the dependencies upfront like this.

picklelo commented on pynecone/var.py:
```diff
@@ -801,6 +806,84 @@ def name(self) -> str:
+            return set()
+        self_name = obj.__code__.co_varnames[0]
+        self_is_top_of_stack = False
+        for instruction in dis.get_instructions(obj):
+            if instruction.opname == "LOAD_FAST" and instruction.argval == self_name:
+                self_is_top_of_stack = True
+                continue
+            if self_is_top_of_stack and instruction.opname == "LOAD_ATTR":
+                d.add(instruction.argval)
+            elif self_is_top_of_stack and instruction.opname == "LOAD_METHOD":
+                d.update(self.deps(obj=getattr(self.__objclass__, instruction.argval)))
+            self_is_top_of_stack = False
```
Thanks for sharing will read more about it

ElijahAhianyo commented:
@masenf This  PR seems to error with this code:

```python
def dynamic_index():
    return pc.text("hello")


# Add state and page to the app.
app = pc.App(state=pc.State)
app.add_page(catchall_index, route="path/[var_arg]")
app.compile()

```


```shell
  File "/Users/eli/Documents/work/Pynecone-io/pynecone-forks/elijah/pynecone/pynecone/compiler/compiler.py", line 205, in compile_page
    code = _compile_page(component, state)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/eli/Documents/work/Pynecone-io/pynecone-forks/elijah/pynecone/pynecone/compiler/compiler.py", line 81, in _compile_page
    state=utils.compile_state(state),
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/eli/Documents/work/Pynecone-io/pynecone-forks/elijah/pynecone/pynecone/compiler/utils.py", line 104, in compile_state
    initial_state = state().dict()
                    ^^^^^^^
  File "/Users/eli/Documents/work/Pynecone-io/pynecone-forks/elijah/pynecone/pynecone/state.py", line 103, in __init__
    for var in cvar.deps():
               ^^^^^^^^^^^
  File "/Users/eli/Documents/work/Pynecone-io/pynecone-forks/elijah/pynecone/pynecone/var.py", line 873, in deps
    d.update(self.deps(obj=getattr(self.__objclass__, instruction.argval)))
                                   ^^^^^^^^^^^^^^^^^
AttributeError: 'ComputedVar' object has no attribute '__objclass__'. Did you mean: '__class__'?

```

masenf commented:
@ElijahAhianyo  thanks for the report, will fix up promptly

masenf commented:
Issue has been identified. Fix is simple, just doing some due diligence to add a test case and see if there are other parts of the code base that add ComputedVar directly to a state class.

masenf commented:
@ElijahAhianyo fix is [here](https://github.com/pynecone-io/pynecone/pull/952)
